%% Generated by Sphinx.
\def\sphinxdocclass{report}
\documentclass[letterpaper,10pt,english]{sphinxmanual}
\ifdefined\pdfpxdimen
   \let\sphinxpxdimen\pdfpxdimen\else\newdimen\sphinxpxdimen
\fi \sphinxpxdimen=.75bp\relax

\usepackage[utf8]{inputenc}
\ifdefined\DeclareUnicodeCharacter
 \ifdefined\DeclareUnicodeCharacterAsOptional\else
  \DeclareUnicodeCharacter{00A0}{\nobreakspace}
\fi\fi
\usepackage{cmap}
\usepackage[T1]{fontenc}
\usepackage{amsmath,amssymb,amstext}
\usepackage{babel}
\usepackage{times}
\usepackage[Bjarne]{fncychap}
\usepackage[dontkeepoldnames]{sphinx}

\usepackage{geometry}

% Include hyperref last.
\usepackage{hyperref}
% Fix anchor placement for figures with captions.
\usepackage{hypcap}% it must be loaded after hyperref.
% Set up styles of URL: it should be placed after hyperref.
\urlstyle{same}
\addto\captionsenglish{\renewcommand{\contentsname}{Contents:}}

\addto\captionsenglish{\renewcommand{\figurename}{Fig.}}
\addto\captionsenglish{\renewcommand{\tablename}{Table}}
\addto\captionsenglish{\renewcommand{\literalblockname}{Listing}}

\addto\extrasenglish{\def\pageautorefname{page}}

\setcounter{tocdepth}{1}



\title{shear Documentation}
\date{Jun 20, 2017}
\release{2017-06-18}
\author{James B. Pease}
\newcommand{\sphinxlogo}{\vbox{}}
\renewcommand{\releasename}{Release}
\makeindex

\begin{document}

\maketitle
\sphinxtableofcontents
\phantomsection\label{\detokenize{index::doc}}



\chapter{Getting Started}
\label{\detokenize{intro:getting-started}}\label{\detokenize{intro:welcome-to-shear-s-documentation}}\label{\detokenize{intro::doc}}\label{\detokenize{intro:intro}}

\section{What is SHEAR?}
\label{\detokenize{intro:what-is-shear}}
SHEAR (Simply Handler for Error and Adapter Removal) is a short-read trimmer for high-throughput sequencing fastq files.
SHEAR first scans the fastq file(s) and automatically detects likely adapter and primer contaminants.  Then it calls
Scythe (\sphinxurl{https://github.com/vsbuffalo/scythe}), which removes reads using a Bayesian error-tolerant approach that
effectively removes adapters even if the adapter itself contains a sequence variation due to sequencing error.
SHEAR then trims and filters reads based on minimum quality and content cutoffs.  Additionally, SHEAR is designed to
automate the simultaneous trimming and concatenation of multiple paired read files into a single trimmed file ready
for mapping or assembly.


\section{Requirements}
\label{\detokenize{intro:requirements}}\begin{itemize}
\item {} 
Python 2.7.x or 3.x (3.x recommended)

\end{itemize}


\subsection{Optional}
\label{\detokenize{intro:optional}}\begin{itemize}
\item {} 
Scythe: \sphinxurl{https://github.com/vsbuffalo/scythe} (\sphinxstylestrong{Strongly recommended})

\end{itemize}


\section{Installation}
\label{\detokenize{intro:installation}}
No installation is necessary, simply clone the repository from GitHub. Scythe should be installed according to its instructions.:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{git} \PYG{n}{clone} \PYG{n}{https}\PYG{p}{:}\PYG{o}{/}\PYG{o}{/}\PYG{n}{www}\PYG{o}{.}\PYG{n}{github}\PYG{o}{.}\PYG{n}{com}\PYG{o}{/}\PYG{n}{jbpease}\PYG{o}{/}\PYG{n}{shear}
\end{sphinxVerbatim}


\section{Preparing your data}
\label{\detokenize{intro:preparing-your-data}}
Standard fastq files with four lines per entry (header, sequence, gap, quality) should be used.  ABI solid colorspace reads are not currently supported.  When using paired-end mode reads only need to be sorted and contain the same number of reads if the \sphinxcode{-U/-{-}filter-unpaired} mode is used, since this will remove both reads from a pair when either of them is filtered out.


\section{Usage}
\label{\detokenize{intro:usage}}

\subsection{Paired-end}
\label{\detokenize{intro:paired-end}}\begin{quote}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{python} \PYG{n}{shear}\PYG{o}{.}\PYG{n}{py} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{fq1} \PYG{n}{FASTQ}\PYG{o}{.}\PYG{n}{SAMPLE1}\PYG{o}{.}\PYG{n}{p1}\PYG{o}{.}\PYG{n}{fastq} \PYG{n}{FASTQ}\PYG{o}{.}\PYG{n}{SAMPLE2}\PYG{o}{.}\PYG{n}{p1}\PYG{o}{.}\PYG{n}{fastq} \PYG{o}{.}\PYG{o}{.}\PYG{o}{.} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{fq2} \PYG{n}{FASTQ}\PYG{o}{.}\PYG{n}{SAMPLE1}\PYG{o}{.}\PYG{n}{p2}\PYG{o}{.}\PYG{n}{fastq} \PYG{n}{FASTQ}\PYG{o}{.}\PYG{n}{SAMPLE2}\PYG{o}{.}\PYG{n}{p2}\PYG{o}{.}\PYG{n}{fastq} \PYG{o}{.}\PYG{o}{.}\PYG{o}{.}  \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{out1} \PYG{n}{FASTQ}\PYG{o}{.}\PYG{n}{sheared}\PYG{o}{.}\PYG{n}{p1}\PYG{o}{.}\PYG{n}{fq} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{out2} \PYG{n}{FASTQ}\PYG{o}{.}\PYG{n}{sheared}\PYG{o}{.}\PYG{n}{p2}\PYG{o}{.}\PYG{n}{fq}
\end{sphinxVerbatim}
\end{quote}


\subsection{Single-read}
\label{\detokenize{intro:single-read}}\begin{quote}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{python} \PYG{n}{shear}\PYG{o}{.}\PYG{n}{py} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{fq1} \PYG{n}{FASTQ}\PYG{o}{.}\PYG{n}{SAMPLE1}\PYG{o}{.}\PYG{n}{p1}\PYG{o}{.}\PYG{n}{fastq} \PYG{n}{FASTQ}\PYG{o}{.}\PYG{n}{SAMPLE2}\PYG{o}{.}\PYG{n}{p1}\PYG{o}{.}\PYG{n}{fastq} \PYG{o}{.}\PYG{o}{.}\PYG{o}{.} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{fq2} \PYG{n}{FASTQ}\PYG{o}{.}\PYG{n}{SAMPLE1}\PYG{o}{.}\PYG{n}{p2}\PYG{o}{.}\PYG{n}{fastq} \PYG{n}{FASTQ}\PYG{o}{.}\PYG{n}{SAMPLE2}\PYG{o}{.}\PYG{n}{p2}\PYG{o}{.}\PYG{n}{fastq} \PYG{o}{.}\PYG{o}{.}\PYG{o}{.}  \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{out1} \PYG{n}{FASTQ}\PYG{o}{.}\PYG{n}{sheared}\PYG{o}{.}\PYG{n}{p1}\PYG{o}{.}\PYG{n}{fq} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{out2} \PYG{n}{FASTQ}\PYG{o}{.}\PYG{n}{sheared}\PYG{o}{.}\PYG{n}{p2}\PYG{o}{.}\PYG{n}{fq}
\end{sphinxVerbatim}
\end{quote}


\subsection{Config file alternative}
\label{\detokenize{intro:config-file-alternative}}
Alternatively to a full set of command line arguments you can enter a single positional argument that points in a text file.  You can then specify command line arguments more neatly over several lines as in the example:
\begin{quote}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{fq1}
\PYG{n}{FASTQ}\PYG{o}{.}\PYG{n}{SAMPLE1}\PYG{o}{.}\PYG{n}{p1}\PYG{o}{.}\PYG{n}{fastq}
\PYG{n}{FASTQ}\PYG{o}{.}\PYG{n}{SAMPLE2}\PYG{o}{.}\PYG{n}{p1}\PYG{o}{.}\PYG{n}{fastq}
\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{fq2}
\PYG{n}{FASTQ}\PYG{o}{.}\PYG{n}{SAMPLE1}\PYG{o}{.}\PYG{n}{p2}\PYG{o}{.}\PYG{n}{fastq}
\PYG{n}{FASTQ}\PYG{o}{.}\PYG{n}{SAMPLE2}\PYG{o}{.}\PYG{n}{p2}\PYG{o}{.}\PYG{n}{fastq}
\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{out1} \PYG{n}{FASTQ}\PYG{o}{.}\PYG{n}{sheared}\PYG{o}{.}\PYG{n}{p1}\PYG{o}{.}\PYG{n}{fq}
\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{out2} \PYG{n}{FASTQ}\PYG{o}{.}\PYG{n}{sheared}\PYG{o}{.}\PYG{n}{p2}\PYG{o}{.}\PYG{n}{fq}
\end{sphinxVerbatim}
\end{quote}


\chapter{Program Parameter Descriptions}
\label{\detokenize{prog_desc:program-parameter-descriptions}}\label{\detokenize{prog_desc::doc}}

\section{adapt}
\label{\detokenize{prog_desc:adapt}}

\subsection{Description}
\label{\detokenize{prog_desc:description}}
This progam (as part of SHEAR) searches adapter sequences
and generates an adapter file for use with SHEAR/Scythe.


\subsection{Parameters}
\label{\detokenize{prog_desc:parameters}}

\subsubsection{\sphinxstyleliteralintitle{-h/-{-}help}}
\label{\detokenize{prog_desc:h-help}}
\sphinxstylestrong{Description:} show this help message and exit

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-{-}fq1} (required)}
\label{\detokenize{prog_desc:fq1-required}}
\sphinxstylestrong{Description:} one or more fastq file paths, separated by spaces

\sphinxstylestrong{Type:} file path; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-o/-{-}out} (required)}
\label{\detokenize{prog_desc:o-out-required}}
\sphinxstylestrong{Description:} output FASTA of adapters detected

\sphinxstylestrong{Type:} file path; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}fq2}}
\label{\detokenize{prog_desc:fq2}}
\sphinxstylestrong{Description:} one or more fastq file paths separated by spaces, only use this for paired-end fastq files and enter these files in the same order as their counterparts in \textendash{}fq1

\sphinxstylestrong{Type:} file path; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-k/-{-}end-klength}}
\label{\detokenize{prog_desc:k-end-klength}}
\sphinxstylestrong{Description:} Length of end kmer to tabulate for possible adapter matches.

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} 16


\subsubsection{\sphinxstyleliteralintitle{-m/-{-}mode}}
\label{\detokenize{prog_desc:m-mode}}
\sphinxstylestrong{Description:} known=only use list of known adapters;endmer=search for common 3’end sequences;both=both known and endmers

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} known

\sphinxstylestrong{Choices:} (‘known’, ‘endmer’, ‘both’)


\subsubsection{\sphinxstyleliteralintitle{-{-}quiet}}
\label{\detokenize{prog_desc:quiet}}
\sphinxstylestrong{Description:} Suppress progress messages

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-M/-{-}min-match}}
\label{\detokenize{prog_desc:m-min-match}}
\sphinxstylestrong{Description:} Minimum proportion of read match required to report the kmer as a possible match.

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} 0.0001


\subsubsection{\sphinxstyleliteralintitle{-N/-{-}number-of-reads}}
\label{\detokenize{prog_desc:n-number-of-reads}}
\sphinxstylestrong{Description:} Number of reads to search in each fastq

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} 200000


\section{shear}
\label{\detokenize{prog_desc:shear}}

\subsection{Description}
\label{\detokenize{prog_desc:id1}}
SHEAR is a read trimmer that coordinates the automatic
finding of adapter sequences, removes adapters using Scythe,
implements various trimming and filtering options
for high-throughput short read sequences, and allows coordinated
removal of paired end sequence files.


\subsection{Parameters}
\label{\detokenize{prog_desc:id2}}

\subsubsection{\sphinxstyleliteralintitle{-h/-{-}help}}
\label{\detokenize{prog_desc:id3}}
\sphinxstylestrong{Description:} show this help message and exit

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-{-}fq1} (required)}
\label{\detokenize{prog_desc:id4}}
\sphinxstylestrong{Description:} one or more fastq file paths, separated by spaces

\sphinxstylestrong{Type:} file path; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}out1} (required)}
\label{\detokenize{prog_desc:out1-required}}
\sphinxstylestrong{Description:} Output fastq file path. Note this is a single output file that concatenates the processed outputs from all files in \textendash{}fq1

\sphinxstylestrong{Type:} file path; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-a/-{-}adapters}}
\label{\detokenize{prog_desc:a-adapters}}
\sphinxstylestrong{Description:} Skip adapter finding an use these adapter files. Either enter (1) a single  file to use for all fastq files, (2)  one file per single end file or pair  of paired-end files. Adapter file(s) should be in FASTA format.

\sphinxstylestrong{Type:} file path; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}clean-header}}
\label{\detokenize{prog_desc:clean-header}}
\sphinxstylestrong{Description:} removes any additional terms from theheader line (useful after STAR)

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-f/-{-}trim-fixed}}
\label{\detokenize{prog_desc:f-trim-fixed}}
\sphinxstylestrong{Description:} Trim a fixed number of bases from the FRONT:END of the sequence (NOT recommended).

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} 0:0


\subsubsection{\sphinxstyleliteralintitle{-{-}filt1}}
\label{\detokenize{prog_desc:filt1}}
\sphinxstylestrong{Description:} Output fastq file path for sequences that werefiltered outs. Note this is a single output file that concatenates the rejectedoutputs from all files in \textendash{}fq1.Default is {[}\textendash{}out1{]}\_filtered\_1.fastq

\sphinxstylestrong{Type:} file path; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}filt2}}
\label{\detokenize{prog_desc:filt2}}
\sphinxstylestrong{Description:} Output fastq file path for sequences that were filtered out. Note this is a single output file that concatenates the rejected outputs from all files in \textendash{}fq2. Default is {[}\textendash{}out2{]}\_filtered\_2.fastq

\sphinxstylestrong{Type:} file path; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}fq2}}
\label{\detokenize{prog_desc:id5}}
\sphinxstylestrong{Description:} one or more fastq file paths separated by spaces, only use this for paired-end fastq files and enter these files in the same order as their counterparts in \textendash{}fq1

\sphinxstylestrong{Type:} file path; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-k/-{-}adapter-end-klength}}
\label{\detokenize{prog_desc:k-adapter-end-klength}}
\sphinxstylestrong{Description:} (Adapter finding) Length of end kmer to tabulate for possible adapter matches.

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} 16


\subsubsection{\sphinxstyleliteralintitle{-{-}log-path}}
\label{\detokenize{prog_desc:log-path}}
\sphinxstylestrong{Description:} Manually specify log file path, default is ‘shear\_TIMESTAMP’

\sphinxstylestrong{Type:} file path; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-m/-{-}adapter-mode}}
\label{\detokenize{prog_desc:m-adapter-mode}}
\sphinxstylestrong{Description:} (Adapter finding) known=only use list of known adapters;endmer=search for common 3’end sequences;both=both known and endmers

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} known

\sphinxstylestrong{Choices:} (‘known’, ‘endmer’, ‘both’)


\subsubsection{\sphinxstyleliteralintitle{-n/-{-}retain-ambig}}
\label{\detokenize{prog_desc:n-retain-ambig}}
\sphinxstylestrong{Description:} By default ambiguous nucleotides (N) are removed from both ends of each read. If this flag is specified, N’s are retained.

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-{-}out2}}
\label{\detokenize{prog_desc:out2}}
\sphinxstylestrong{Description:} output fastq file path. Note this is a single output file that concatenates the processed outputs from all files in \textendash{}fq2

\sphinxstylestrong{Type:} file path; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-p/-{-}trim-poly}}
\label{\detokenize{prog_desc:p-trim-poly}}
\sphinxstylestrong{Description:} Trim poly-A or poly-T repeats of at least this length from the front or end.

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} 12


\subsubsection{\sphinxstyleliteralintitle{-q/-{-}trim-qual}}
\label{\detokenize{prog_desc:q-trim-qual}}
\sphinxstylestrong{Description:} Trim bases below this quality score from the FRONT:END of each read.

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} 20:20


\subsubsection{\sphinxstyleliteralintitle{-{-}quality-scale}}
\label{\detokenize{prog_desc:quality-scale}}
\sphinxstylestrong{Description:} Quality scale is usually automatically determined, but use this to set manually.

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None

\sphinxstylestrong{Choices:} (‘sanger’, ‘illumina’, ‘phred’, ‘solexa’)


\subsubsection{\sphinxstyleliteralintitle{-{-}quiet}}
\label{\detokenize{prog_desc:id6}}
\sphinxstylestrong{Description:} Suppress progress messages

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-{-}retain-temp}}
\label{\detokenize{prog_desc:retain-temp}}
\sphinxstylestrong{Description:} Retain temporary files (none=remove all; tempfastq=remove temporary fastq files from scythe; exceptadapters=remove temp fastq and log files from Scythe but keep adapter file; all=keep all temporary file)

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} none

\sphinxstylestrong{Choices:} {[}‘none’, ‘tempfastq’, ‘exceptadapters’, ‘all’{]}


\subsubsection{\sphinxstyleliteralintitle{-s/-{-}scythe-prior}}
\label{\detokenize{prog_desc:s-scythe-prior}}
\sphinxstylestrong{Description:} Bayesian prior for proportion of adapters expected to be sampled in Scythe.

\sphinxstylestrong{Type:} float; \sphinxstylestrong{Default:} 0.1


\subsubsection{\sphinxstyleliteralintitle{-{-}scythe-match}}
\label{\detokenize{prog_desc:scythe-match}}
\sphinxstylestrong{Description:} Minimum number of bases required for a match in Scythe.

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} 5


\subsubsection{\sphinxstyleliteralintitle{-{-}scythe-skip}}
\label{\detokenize{prog_desc:scythe-skip}}
\sphinxstylestrong{Description:} Skip scythe 3’ adapter removal.

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-t/-{-}platform}}
\label{\detokenize{prog_desc:t-platform}}
\sphinxstylestrong{Description:} Sequencing Platform

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} TruSeq

\sphinxstylestrong{Choices:} (‘TruSeq’, ‘TruSeqDualIndex’)


\subsubsection{\sphinxstyleliteralintitle{-{-}temp-dir}}
\label{\detokenize{prog_desc:temp-dir}}
\sphinxstylestrong{Description:} directory to use for temporary files

\sphinxstylestrong{Type:} file path; \sphinxstylestrong{Default:} .


\subsubsection{\sphinxstyleliteralintitle{-{-}trim-qual-pad}}
\label{\detokenize{prog_desc:trim-qual-pad}}
\sphinxstylestrong{Description:} Trim additional bases next to low-quality bases specified by \textendash{}trim-qual from the FRONT:END of each read.

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} 0:0


\subsubsection{\sphinxstyleliteralintitle{-y/-{-}trim-pattern-5}}
\label{\detokenize{prog_desc:y-trim-pattern-5}}
\sphinxstylestrong{Description:} Comma-separated list of specific sequences to trim from the 5’ end. (Not recommmended).

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-z/-{-}trim-pattern-3}}
\label{\detokenize{prog_desc:z-trim-pattern-3}}
\sphinxstylestrong{Description:} Comma-separated list of specific sequences to trim from the 3’ end. Can be used for extra stringent adapter trimming.

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-A/-{-}filter-ambig}}
\label{\detokenize{prog_desc:a-filter-ambig}}
\sphinxstylestrong{Description:} Filter reads with more than this numberof ambiguous nucleotides (N’s; set as 0 to skip

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} 5


\subsubsection{\sphinxstyleliteralintitle{-I/-{-}filter-low-info}}
\label{\detokenize{prog_desc:i-filter-low-info}}
\sphinxstylestrong{Description:} Filter out reads with mutual information scores exceeding this value (ADVANCED, removes highly repetitive reads).

\sphinxstylestrong{Type:} float; \sphinxstylestrong{Default:} 0.0


\subsubsection{\sphinxstyleliteralintitle{-L/-{-}filter-length}}
\label{\detokenize{prog_desc:l-filter-length}}
\sphinxstylestrong{Description:} Filter out reads that contain fewer than this many characters after trimming.

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} 30


\subsubsection{\sphinxstyleliteralintitle{-M/-{-}adapter-min-match}}
\label{\detokenize{prog_desc:m-adapter-min-match}}
\sphinxstylestrong{Description:} (Adapter finding) Minimum proportion of read match required to report the kmer as a possible match.

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} 0.0001


\subsubsection{\sphinxstyleliteralintitle{-N/-{-}adapter-number-of-reads}}
\label{\detokenize{prog_desc:n-adapter-number-of-reads}}
\sphinxstylestrong{Description:} (Adapter finding) Number of reads to search in each fastq

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} 200000


\subsubsection{\sphinxstyleliteralintitle{-Q/-{-}filter-quality}}
\label{\detokenize{prog_desc:q-filter-quality}}
\sphinxstylestrong{Description:} Filter out reads with a mean quality score below this value (before trimming).

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} 3


\subsubsection{\sphinxstyleliteralintitle{-U/-{-}filter-unpaired}}
\label{\detokenize{prog_desc:u-filter-unpaired}}
\sphinxstylestrong{Description:} If either read in a read pair is filtered out, the counterpart reads is also filtered out regardless of quality.

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-X/-{-}scythe-executable}}
\label{\detokenize{prog_desc:x-scythe-executable}}
\sphinxstylestrong{Description:} Set the path of the scythe executable manually.

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} scythe


\chapter{Indices and tables}
\label{\detokenize{index:indices-and-tables}}\begin{itemize}
\item {} 
\DUrole{xref,std,std-ref}{genindex}

\item {} 
\DUrole{xref,std,std-ref}{modindex}

\item {} 
\DUrole{xref,std,std-ref}{search}

\end{itemize}



\renewcommand{\indexname}{Index}
\printindex
\end{document}